### PR TITLE
kvutils: Remove fingerprints from the pre-executing validator. [KVL-747]

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
The submission validator doesn't care about fingerprints, it only needs to know about them in order to discard them from the state value. This introduces a new typeclass, `HasDamlStateValue`, which can be customized for various state value types to make submission validation generic.

A little more code for now, but we're setting things up so it can be deleted later.

This is part 4 of removing fingerprints from this code, while still maintaining the ability to pass extra ledger-specific information through.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by fabiotudone-da — APPROVED** [review]

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/package.scala [hunk 1]
```diff
@@ -42,4 +43,11 @@ package object kvutils {
 
   implicit val `Bytes Ordering`: Ordering[Bytes] = Ordering.by(_.asReadOnlyByteBuffer)
 
+  implicit object `DamlStateValue with Fingerprint has DamlStateValue`
+      extends HasDamlStateValue[(Option[DamlStateValue], Fingerprint)] {
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/package.scala:47]: Would it make sense to add a type alias for this to improve readability?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala [hunk 2]
```diff
@@ -30,22 +19,22 @@ import scala.concurrent.{ExecutionContext, Future}
 /**
   * Validator for pre-executing submissions.
   */
-class PreExecutingSubmissionValidator[ReadSet, WriteSet](
+class PreExecutingSubmissionValidator[StateValue, ReadSet, WriteSet](
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala:22]: I'd prefer if this type parameter implied that it is wrapping a DAML state value. E.g., a more specific name may be `DecoratedDamlStateValue`. Alternatively, please add some code docs.

**Review by miklos-da — APPROVED** [review]

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/package.scala (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/package.scala:47]: Gonna delete it shortly, so no thanks. :upside_down_face:

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala]: I've added some docs to explain. Thank you for the nudge to do so. :grinning:

**Pull request merged**